nucypher/network/server.py

```diff
@@ -156,30 +151,40 @@ def node_metadata_exchange():
 
     @rest_app.route('/reencrypt', methods=["POST"])
     def reencrypt():
-
+        # TODO: Cache & Optimize
         from nucypher.characters.lawful import Bob
 
-        # TODO: Cache & Optimize
+        # Deserialize and instantiate the request
         reenc_request = ReencryptionRequest.from_bytes(request.data)
 
+        # Deserialize and instantiate ConditionLingo from the request data
         json_lingo = json.loads(str(reenc_request.conditions))
         lingo = [ConditionLingo.from_list(lingo) if lingo else None for lingo in json_lingo]
-        context = json.loads(str(reenc_request.context)) or dict()  # requester-supplied input
+
+        # requester-supplied reencryption condition context
+        context = json.loads(str(reenc_request.context)) or dict()
+
+        # zip capsules with their respective conditions
         packets = zip(reenc_request.capsules, lingo)
 
+        # Populate default request context for decentralized nodes
+        if not this_node.federated_only:
+            context.update({'providers': this_node.condition_providers})
```

Comment:
Any reason we shouldn't just pass `condition_providers` directly to `evaluate_conditions_for_ursula(lingo, context, log, ursula)` and subsequently `lingo.eval(...)` method instead of piggybacking ontop of `context`? Perhaps this was the original reason for the unused `ursula` parameter in `evaluate_conditions_for_ursula`?

I'm cautious of modifying a `context` object that was obtained directly via a user request. I like the idea of passing `condition_providers` and `context` to the `eval(...)` method (similar to `verify(...)`); feels like a good separation.

Reply:
I agree.  Removed this usage of context (Ursula no longer mutates user context).  In retrospect I was looking for the most elegant way support federated mode and settled on this imperfect solution.  Instead I added ternary operation to handle providers:

```python
providers = this_node.condition_provider if not this_node.federated_only else dict()
```

In the highly unlikely situation where an onchain condition is submitted to evaluation by an ursula, the empty dictionary here will fire this `KeyError` in `nucypher/policy/conditions/evm.py`:

```python
def _configure_provider(self, providers: Dict[int, BaseProvider]):
    """Binds the condition's contract function to a blockchian provider for evaluation"""
    try:
        provider = providers[self.chain]
    except KeyError:
        raise Exception(f'This node does not have a connection to chain {self.chain}')

```